### DISCUSSION TITLE
Optimistic concurrency violation with migration to 5.X

### DISCUSSION THREAD

**Discussion by mfonnesbeck** [ravendb/ravendb#16779] [Q&A]
We recently started upgrading our Raven Client APIs from 3.x to 5.x and found a Optimistic concurrency violation when saving stored tokens in the database. We use an Accounting Keys object for each tenant that stores the access and refresh tokens for OAuth authentication to our accounting integration partner software. In our 3.x version when the tokens needed to be refreshed, in order to avoid concurrency issues, we would use a read lock to load the Accounting Keys object from the database refresh the tokens and then use a write lock to save the changed object back to the database.  We were also retrieving ETags for the object to ensure that it was the same instance of the object stored by turning on UseOptimisticConcurrency on the session before we write the object to the database. The document sessions were obtained using Dependency Injection to get a new session when needed. We did this in order to ensure that only one person is updating and refreshing the tokens at a time.

With the upgrade we are encountering an error when saving to the database:

"Document AccountingKeys-481 has change vector A:5833-LxE+fvEr+EirswfyKiiWGw, but Put was called with expecting new document. Optimistic concurrency violation, transaction will be aborted."

Do we need to worry about the concurrency of objects where the database is tracking things with the Change Vector? I don't fully understand how Raven is using this field. Should we be setting the UseOptimisticConcurrency field on the document session? Please enlighten me. Thank you.

**Comment by ayende** [comment]
Between 3.5 and 5.4, we changed things to use the change vectors, not the etag, for optimistic concurrency.

You are expected to save the change vector of the entity and call back with the same change vector on change.


```
using (var session1 = store.OpenSession())
{
    var document1 = session1.Load<MyDocument>("documents/1");
    var changeVector1 = session1.Advanced.GetChangeVectorFor(document1);

   SendToClient(document1, changeVector1);
}

using (var session2 = store.OpenSession())
{
   var (doc, changeVector) = ReadFromClient();

   session2.Store(doc, changeVector);
   session2.SaveChanges();
}




```

**Reply by mfonnesbeck** [reply]
Thank you for your help. With this object we load it from the database and then get the change vector and store that with the object while we are using it. When we save, we use the change vector with the Store( ) method to ensure concurrency checks. I have a Unit Test that mimics our dependency injector to serve up a document session when the object is loaded and saved. In the Unit Test it is throwing an Optimistic concurrency violation when we try to save a second instance the was saved after the first was saved - which is what it is supposed to do. However, after the exception, I load the object from the database with the object's Id in order to get the current version of the object and return to the calling environment, and in the test (and even when connected to an actual database instance) it loads the old version of the object that we were just trying to save. Is there something cached in memory where it is not loading the saved object?

**Reply by ayende** [reply]
Can you show the unit test?
